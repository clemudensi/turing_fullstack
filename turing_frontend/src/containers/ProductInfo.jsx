import React, {useState, useEffect} from 'react';
import { Grid } from 'semantic-ui-react';
import _ from 'lodash';
import { connect } from 'react-redux';
import ProductImages from '../components/product/ProductImages';
import ProductHeader from '../components/product/ProductHeader';
import ProductColor from '../components/product/ProductColor';
import ProductSize from '../components/product/ProductSize';
import ProductDetail from '../components/product/ProductDetail';
import { fetchProduct } from '../actions/products';
import {IMAGE, PRODUCTS} from '../constants';
import ProductQuantity from '../components/product/ProductQuantity';
import axios from 'axios';
import PageLoader from '../util/PageLoader';


const ProductInfo = (props) => {
  let { cart } = props;
  const { fetchProduct, getProductId, handleAddOnceToCart } = props;
  const [selectedColor, setSelectedColor] = useState('Please Select Color');
  const [selectedSize, setSelectedSize] = useState('Please Select Size');
  const [notFound, setNotFound] = useState(false);
  const [cartId, setCartId] = useState('');

  const handleSelectColor = (value) =>{
    value = value.replace(/^[a-z]?/, color => color.toUpperCase());
    setSelectedColor(value);
  };

  const handleSelectSize = (value) =>{
    setSelectedSize(value.toUpperCase());
  };

  const {
    handleAddToCart,
    quantity,
    handleChange,
    history,
    products: { data },
    handleBackToProducts,
    canAddToCart,
  } = props;
  const { attribute_values } = data;

  const size = _.map(attribute_values, color => color.value.toLowerCase());
  size.splice(size.indexOf('indigo'), 1, 'pink');
  size.splice(size.indexOf('white'), 1, 'grey');
  const color = size.splice(5, 9);

  const dataProInfo = {
    id: data.product_id,
    name: data.name,
    price: data.price,
    salePrice: data.discounted_price,
    image: `${IMAGE}/${data.image}`,
    quantity: quantity,
    customer_id: localStorage.getItem('user_id'),
    color,
    size,
    des: data.description,
    reference: cartId
  };

  const proInfo = {...dataProInfo};
  Object.assign(proInfo, {color: selectedColor, size: selectedSize});

  let hasSelected = false;

  if (selectedColor !== 'Please Select Color' && selectedSize !== 'Please Select Size') {
    hasSelected = true;
  }

  const loading = () => {
    return notFound === false ? <PageLoader/> : <h2>Product not found</h2>;
  };

  useEffect(() => {
    fetchProduct(getProductId());
    const runEffect = async () => {
      const { data: { cart_id } } = await axios.get(`${PRODUCTS}/shoppingcart/generateUniqueId`);
      setCartId(cart_id);
    };
    runEffect();
    if (data === {} ){
      setNotFound(true)
    }
    handleAddOnceToCart();
  }, [cart]);

  return (
    data && cartId ?
      <Grid.Row>
        <Grid.Column width={6}>
          <ProductImages img={proInfo.image} />
        </Grid.Column>
        <Grid.Column id="product-info" width={6} textAlign="left">
          <ProductHeader
            name={dataProInfo.name}
            price={dataProInfo.price}
            salePrice={dataProInfo.salePrice}
          />
          <ProductColor
            dataColor={dataProInfo.color}
            selectedColor={selectedColor}
            handleSelectColor={handleSelectColor}
          />
          <ProductSize
            dataSize={dataProInfo.size}
            selectedSize={selectedSize}
            handleSelectSize={handleSelectSize}
          />
          <ProductQuantity
            handleChange={handleChange}
            quantity={quantity}
          />
          <ProductDetail
            dataDes={dataProInfo.des}
            canAddToCart={canAddToCart}
            hasSelected={hasSelected}
            handleAddToCart={handleAddToCart}
            handleAddOnceToCart={handleAddOnceToCart}
            proInfo={proInfo}
            cart={cart}
            history={history}
            cartId={cartId}
            handleBackToProducts={handleBackToProducts}
          />
        </Grid.Column>
      </Grid.Row> :
      loading()
  );
};

ProductInfo.defaultProps = {
  product: {
    data: false
  },
  location: {
    pathname: ''
  }
};


export default connect(
  ({ products }) => ({ products }),
  { fetchProduct })(ProductInfo);
