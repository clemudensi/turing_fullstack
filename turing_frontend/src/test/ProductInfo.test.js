import React from 'react';
import {Provider} from 'react-redux';
import {mount} from '../enzyme';
import configureMockStore from 'redux-mock-store';
import ProductInfo from '../containers/ProductInfo';

const mockStore = configureMockStore();
const store = mockStore({});


describe('Renders ProductInfo Component', () => {

  const wrapper = mount(<ProductInfo store={store} />);
  it('Should render productInfo', () => {
    expect(wrapper.find('#product-info')).toBeDefined();
    expect(wrapper.find('#product-info').length).toBeGreaterThan(0);
    expect(wrapper.find('ProductHeader')).toHaveLength(1);
    expect(wrapper.find('ProductSize')).toHaveLength(1);
    expect(wrapper.find('ProductColor')).toHaveLength(1);
    expect(wrapper.find('ProductQuantity')).toHaveLength(1);
    expect(wrapper.find('ProductDetail')).toHaveLength(1);
    expect(wrapper.find('#product-info').exists()).toEqual(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render ProductHeader with props', () => {
    expect(wrapper.find('ProductHeader').props()).toHaveProperty('salePrice');
    expect(wrapper.find('ProductHeader').props()).toHaveProperty('price');
    expect(wrapper.find('ProductHeader').props()).toHaveProperty('name');
  });

  it('should render ProductSize with props', () => {
    expect(wrapper.find('ProductSize').props()).toHaveProperty('dataSize');
    expect(wrapper.find('ProductSize').props()).toHaveProperty('selectedSize');
    expect(wrapper.find('ProductSize').props()).toHaveProperty('handleSelectSize');
  });

  it('should render ProductColor with props', () => {
    expect(wrapper.find('ProductColor').props()).toHaveProperty('dataColor');
    expect(wrapper.find('ProductColor').props()).toHaveProperty('selectedColor');
    expect(wrapper.find('ProductColor').props()).toHaveProperty('handleSelectColor');
  });

  it('should render ProductQuantity with props', () => {
    expect(wrapper.find('ProductQuantity').props()).toHaveProperty('handleChange');
    expect(wrapper.find('ProductQuantity').props()).toHaveProperty('quantity');
  });

  it('should render ProductDetail with props', () => {
    expect(wrapper.find('ProductDetail').props()).toHaveProperty('dataDes');
    expect(wrapper.find('ProductDetail').props()).toHaveProperty('canAddToCart');
    expect(wrapper.find('ProductDetail').props()).toHaveProperty('hasSelected');
    expect(wrapper.find('ProductDetail').props()).toHaveProperty('handleAddToCart');
    expect(wrapper.find('ProductDetail').props()).toHaveProperty('handleAddOnceToCart');
    expect(wrapper.find('ProductDetail').props()).toHaveProperty('proInfo');
    expect(wrapper.find('ProductDetail').props()).toHaveProperty('cart');
    expect(wrapper.find('ProductDetail').props()).toHaveProperty('history');
  });
});