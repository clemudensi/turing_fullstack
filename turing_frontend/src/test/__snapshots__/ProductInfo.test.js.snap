// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ProductInfo Component Should render productInfo 1`] = `ReactWrapper {}`;
